### DISCUSSION TITLE
micromark-abbr extension

### DISCUSSION THREAD

**Discussion by richardTowers** [micromark/micromark#181] [Q&A] [answered]
Hello! First of all thank you for micromark - lovely code, really elegant approach to a really inelegant problem space 😅 

I've been working on an extension to support the `abbr` syntax found in some markdown flavours ([PHP markdown extras](https://michelf.ca/projects/php-markdown/extra/#abbr), [kramdown](https://kramdown.gettalong.org/syntax.html#abbreviations)). My code is here - https://github.com/richardTowers/micromark-abbr

## Context

A bit of extra context so I don't do an XY problem.

In my day job, I work on the GOV.UK Publishing team. We have our own extension of markdown called [Govspeak](https://github.com/alphagov/govspeak), which has [numerous syntax extensions](https://govspeak-preview.publishing.service.gov.uk/guide).

Govspeak written in Ruby (based on Kramdown), which means it's not easy to run it in the browser. The implementation is also quite fragile and buggy, relying heavily on pre-processing the markdown with regex replacements.

I'm interested in re-implementing the Govspeak syntax in a more considered way in JavaScript, so that we can have a less buggy and fragile implementation that can be run in web browsers.

micromark / remark is my current preferred option, but I haven't dismissed using markdown-it or another framework yet. I also haven't dismissed giving up completely, but hopefully we won't get there 😅

## Questions

I've got a [minimally working example of an abbr extension](https://github.com/richardTowers/micromark-abbr
) now, which is exciting! Mostly I followed the tutorial in the README and heavily cribbed off the micromark-gfm-footnotes extension.

There are two areas where I had to do things which I wasn't very happy about...

### Thing 1: Starting characters for abbr calls

The abbr syntax is horrible, in that there actually is no syntax for abbr calls. If you have:

```markdown
The HTML specification is maintained by the W3C.

*[HTML]: Hyper Text Markup Language
*[W3C]:  World Wide Web Consortium
```

Then `HTML` and `W3C` should be parsed as `abbrCall`. If those labels aren't defined, then they're just text.

The difficulty here is that you don't know what characters might begin an `abbrCall` expresion until you've parsed the document.

I worked around this by [starting on all uppercase ASCII characters](https://github.com/richardTowers/micromark-abbr/blob/7c76f7289210378c512c14d06473d6808ec5bdd0/dev/lib/syntax.js#L174-L180), but I think this overly restrictive - a fully compliant implementation would work for lowercase letters, and also unicode labels. I did look at having the parser modify itself, so the `text` parser doesn't get defined until we've parsed `document` and know what the labels are, but I couldn't make it work.

Do you have any suggestions on a nicer way to support that?

### Thing 2: Hoisting events

As discussed in https://github.com/orgs/micromark/discussions/78 , `definition` is special, in that the HTML compiler reorders events so that definitions come first. This means that link references will always come after their definitions in the event list, which makes it possible for them to refer back to the data in the definition.

We need similar functionality for abbr definitions, but the syntax isn't _quite_ the same as the syntax for link reference definitions:

```markdown
<!-- A link reference definition -->
[micromark]: github.com/micromark/micromark "The micromark markdown parser / compiler"

<!-- An abbr definition -->
*[HTML]: Hyper Text Markup Language
```

The built in `definition` won't parse abbr definitions, even if we consume the asterisk first, becuase it's expecting a URL-like thing after the colon and we want a bunch of text with whitespaces in it. The quoted title bit is optional, so we can just ignore that.

The way I've worked around this is by [using definition and definitionLabelString states](https://github.com/richardTowers/micromark-abbr/blob/7c76f7289210378c512c14d06473d6808ec5bdd0/dev/lib/syntax.js#L15-L18), even though technically these are a slightly different kind of thing. This means they get hoisted up to the start of the events list by the compiler, and I can use the data in abbr calls.

How ugly is this work around in your view? Is there already a better way to do it?

I guess I could not provide an HTML compiler in the micromark extension, and do the transformation on the AST in a later step in the mdast / remark / rehype chain.

Alternatively, would you consider some change to micromark to allow extension-defined events to be hoisted to the start of the events list?

## End note

Just to be completely open - I'm also happy to hear answers of the form "we don't think it's sensible to build a micromark extension for abbr", or "even if abbr is welcome, some of those other things in your govspeak parser look like they would be too horrible to implement". I can always look at markdown-it and other parsers.

I've really enjoyed playing around writing this extension - it's the first time I've done something with parsers and compilers, and it's been fun.

Thanks!

**Comment by wooorm** [comment, marked as answer]
Hi Richard!

How interesting, Govspeak.

> run in web browsers

What is the reason JS is needed for this? Could not a `/govspeak` endpoint that turns Govspeak into HTML be created?

> I haven't dismissed using markdown-it or another framework yet.

Have you seen https://github.com/micromark/micromark#markdown-it? (deep link).

> giving up completely

We do often recommend folks not do syntax extensions, have you seen https://github.com/micromark/micromark#extending-markdown (deep link).

Would it somehow be possible to switch to, say, a govspeak 2, which uses directives (a singular extension syntax for different extensions?)

***

> and also unicode labels

What do you mean by “unicode labels”? Do you mean that *any* punctuation, whitespace, symbol, letter, number, anything, can be used as a label?

> Do you have any suggestions on a nicer way to support that?

You might be able to ignore this: parse the definitions only. Then, when you have an AST, look for them.

GFM email autolinks (`asd user@example.com qwe`) have a similar problem, but is limited to letters (like your ascii characters).
In Rust—which I made later, sort of like a 2nd iteration on the algorithm, and is in the process of being ported back—I already look for `@` afterwards and then go through the events.
See https://github.com/micromark/micromark-extension-gfm-autolink-literal/blob/14548d89b3cf6ba678d2b29f978d7b4cd4a08206/dev/lib/syntax.js#L77-L80.
You could do something like that, with a “resolver”. Whether `*` turns into emphasis/strong/etc, or `[` and `]…` into images/links, is also done with resolvers. https://github.com/micromark/micromark/blob/4bcb4cc58ff360b2e7db15c1f60995d806b6827f/packages/micromark-core-commonmark/dev/lib/attention.js#L25.

What to go with, depends. Perhaps on the grammar too. Which of these works?

````markdown
# W3C?

W3C? w3c?

`W3C`? *W3C*? **W3C**?

[W3C?](W3C? "W3C?") ![W3C?](W3C? "W3C?") <https://W3C?>

```W3C?
W3C?
```

*[HTML]: Do other abbreviations work? W3C?
*[W3C]: World Wide Web Consortium (and how about recursion? W3C?)
````

***

> How ugly is this work around in your view? Is there already a better way to do it?

Pretty ugly! Likely to break at some point. I’d really recommend unique names for your things
But, for a better way? 🤔

> I guess I could not provide an HTML compiler in the micromark extension, and do the transformation on the AST in a later step in the mdast / remark / rehype chain.

Indeed, that could work. Still, for me, if there’s a `micromark-extension-abbr`, I would like it to work without the ASTs.

> Alternatively, would you consider some change to micromark to allow extension-defined events to be hoisted to the start of the events list?

Right. If that’s needed, that must happen.

This reordering is not needed for footnote definitions though: https://github.com/micromark/micromark-extension-gfm-footnote/blob/0a62fad40470f2447707020c52d38d1494199ee1/dev/lib/html.js#L145 🤔

I do wonder though.
Perhaps there is also alternative HTML that can be generated.
Right now, you are using `title` attributes right?
They have some problems: not guaranteed to show especially on mobile; no support for rich content.
There’s also a lot of reuse: if `W3C` is used 10 times, the title is repeated 10 times.
A few weeks ago I made [`rehype-twoslash`](https://github.com/rehypejs/rehype-twoslash), which generates “popovers” for TypeScript errors. With some CSS, JS, and a `popover` attribute.
Perhaps you can also generate `<span data-id="abbr-w3c">W3C</span>` for every “call”, and then turn the definition, wherever it appears, into the related corresponding markup. You don’t need the definition “higher” with that.

***

> we don't think it's sensible to build a micromark extension for abbr

No, whether it’s recommended is one thing. And there might be really weird syntaxes one comes up with for which that is the answer. But abbreviations, stuff like that, should be possible.

> even if abbr is welcome, some of those other things in your govspeak parser look like they would be too horrible to implement

Ah, right. Ehh. Well, then I need to review them 😅

Steps seems weird.
Buttons seems like you could do something with a fork of `micromark-extension-directive`.
Heading, I recommend suggesting a space after `#`, it doesn’t work without it in CommonMark.

> I've really enjoyed playing around writing this extension - it's the first time I've done something with parsers and compilers, and it's been fun.

Glad to hear! ASTs are already tough for most people. Integrating into compilers even more so. Cool to hear that you enjoy it!

**Reply by richardTowers** [reply]
> > run in web browsers

> What is the reason JS is needed for this? Could not a /govspeak endpoint that turns Govspeak into HTML be created?

Yes, that's more or less what we do at the moment. It makes certain kinds of UX difficult though - for example a side-by-side live preview of the rendered HTML would have quite high latency (and load on our servers). Some govspeak documents can be very long, so we'd probably not be able to provide UX like that for long documents.

There are also some use cases where it would be nice to have the AST on the client - for example, we're working on a  WYSIWYG editor based on prosemirror which should serve as an alternative to markdown. But it doesn't have full feature parity. It would be nice to be able to walk a markdown tree to see if it uses any constructs the WYSIWYG doesn't support to do that toggle.

All that said, being able to run in the browser is still more of a nice to have than an actual requirement. It's just an interesting idea I'm exploring at this point, rather than something we're committed to doing.

> > I haven't dismissed using markdown-it or another framework yet.

> Have you seen https://github.com/micromark/micromark#markdown-it? (deep link).

Yeah - that's what made me think it could be a good alternative. They already have an abbr extension by the looks of it, and I'm not too bothered by the size of it. I'm still interested to see if micromark / remark will work for me though.

> > giving up completely

> We do often recommend folks not do syntax extensions, have you seen https://github.com/micromark/micromark#extending-markdown (deep link).

> Would it somehow be possible to switch to, say, a govspeak 2, which uses directives (a singular extension syntax for different extensions?)

Possible, yes, but practical probably not. We've got about 700k pages on GOV.UK, which are all authored in govspeak. So we'd have to convert all of those. We've also got around 2k users, who are experienced with the current syntax and fairly resistant to change. It's not totally beyond the realms of possibility, but it would be an expensive project.

For the most part govspeak (and its current ruby implementation) is working well for us. It can be difficult to train people on it (hence WYSIWYG), and some things (particularly tables) are a nightmare for our users. But switching to govspeak 2 would probably not fix those issues, so it's unlikely to happen from that angle too.

***

and also unicode labels

> What do you mean by “unicode labels”? Do you mean that any punctuation, whitespace, symbol, letter, number, anything, can be used as a label?

Sorry - badly phrased. I guess I mean that, strictly speaking, people should be able to do things like:

```markdown
I don't speak Russian, but a quick search of russian acronyms brought up БДИПЧ as an example.

*[БДИПЧ]: Бюро по демократическим институтам и правам человека
```

... which wouldn't work if I only kick the tokenizer off on ASCII characters. But would work fine if I could use the definitions.

> > Do you have any suggestions on a nicer way to support that?

> You could do something like that, with a “resolver”

That's a good suggestion - I'll have a go.

> Which of these works?

Let's see what Kramdown does...

````ruby
require 'bundler/inline'

gemfile do
  source 'https://rubygems.org'
  gem 'kramdown'
end

require 'kramdown'

puts Kramdown::Document.new(<<MARKDOWN).to_html
# W3C?

W3C? w3c?

`W3C`? *W3C*? **W3C**?

[W3C?](W3C? "W3C?") ![W3C?](W3C? "W3C?") <https://W3C?>

```W3C?
W3C?
```

*[HTML]: Do other abbreviations work? W3C?
*[W3C]: World Wide Web Consortium (and how about recursion? W3C?)
MARKDOWN
````

````html
<h1 id="w3c"><abbr title="World Wide Web Consortium (and how about recursion? W3C?)">W3C</abbr>?</h1>

<p><abbr title="World Wide Web Consortium (and how about recursion? W3C?)">W3C</abbr>? w3c?</p>

<p><code>W3C</code>? <em><abbr title="World Wide Web Consortium (and how about recursion? W3C?)">W3C</abbr></em>? <strong><abbr title="World Wide Web Consortium (and how about recursion? W3C?)">W3C</abbr></strong>?</p>

<p><a href="W3C?" title="W3C?"><abbr title="World Wide Web Consortium (and how about recursion? W3C?)">W3C</abbr>?</a> <img src="W3C?" alt="W3C?" title="W3C?" /> <a href="https://W3C?">https://<abbr title="World Wide Web Consortium (and how about recursion? W3C?)">W3C</abbr>?</a></p>

<p><code>W3C?
W3C?
</code></p>
````

So it expands in headings, paragraphs, em, strong, anchors. It doesn't expand in code, attributes like src / href / alt / title. It does expand inside the text of `<autolinks>` by the looks of it (but not the href), even though it's inside a URL. I guess that would probably be mostly unexpected behaviour for most users, but then, it's markdown.

***

> > How ugly is this work around in your view? Is there already a better way to do it?

> Pretty ugly! Likely to break at some point. I’d really recommend unique names for your things
But, for a better way? 🤔

Yeah, that's what I thought!

> > Alternatively, would you consider some change to micromark to allow extension-defined events to be hoisted to the start of the events list?

> Right. If that’s needed, that must happen.

I'd be interested in discussing it further. Is this something which we could use a content type for? e.g. give definitions their own content type, then adjust [the code that hoists them](https://github.com/micromark/micromark/blob/4bcb4cc58ff360b2e7db15c1f60995d806b6827f/packages/micromark/dev/lib/compile.js#L299) to hoist anything with a definition-like content type? Or some other way of adding some metadata to the tokens?

I don't think I'd accept a change to the HTML - I'd have to think about accessibility, and users without JS etc. Feels like if one is going to the trouble of adding abbreviations, the abbr tag is the way to go. I do take the point about duplication of titles though.

***

Sounds like an abbr extension is welcome then, if we can do it in a reasonable way. 💪🏻 

Thanks very much for having a glance over those other extensions - you didn't have to do that! I'll have a look at Steps next if that sounds like it could be trouble. Very good point about the space after headings thing - that's an example where I think I'd be happy to drop support for spaceless headings and batch change any existing markdown that's non-compliant.

I imagine I'll find lots of places where Kramdown isn't compliant with CommonMark, so even if I did manage to get a micromark / remark setup that supports all the constructs in govspeak, there would still be a big game of whackamole to fix all of the areas where we're relying on Kramdown not following the spec.

**Reply by wooorm** [reply]
Could you also test:

```markdown
W. X. y. z. 0. 9. A-B. C+D. E0.

*[w]: Whiskey
*[X]: X-ray
*[Y]: Yankee
*[z]: Zulu
*[0]: Zero
*[9]: Nine
*[A-B]: Alpha-bravo
*[C+D]: Charlie delta
*[E0]: Echo zero
```

And:

```markdown
X.

*[X]: *a* **b** `c` D ~~e~~ [f](g)
*[D]: Delta
```

>  It does expand inside the text of `<autolinks>` by the looks of it (but not the href), even though it's inside a URL. I guess that would probably be mostly unexpected behaviour for most users, but then, it's markdown.

This is quite interesting. That would mean it sort of “expands” into `[text](url)` earlier. And works on that.

> I do take the point about duplication of titles though.

I would also recommend looking at the current `*\[`s that are in use currently, to see if anyone has ever put “markdown” in there, like emphasis or links. Are they really only abbreviations? Or are folks using them more like tooltips. Like footnotes.

**Reply by richardTowers** [reply]
They come out as:

````html
<p>W. <abbr title="X-ray">X</abbr>. y. <abbr title="Zulu">z</abbr>. <abbr title="Zero">0</abbr>. <abbr title="Nine">9</abbr>. <abbr title="Alpha-bravo">A-B</abbr>. <abbr title="Charlie delta">C+D</abbr>. <abbr title="Echo zero">E0</abbr>.</p>
````

````html
<p><abbr title="*a* **b** `c` D ~~e~~ [f](g)">X</abbr>.</p>
````

The weirdest abbreviation labels I've been able to find so far are on https://www.gov.uk/guidance/the-highway-code/annex-2-motorcycle-licence-requirements

This has:

```markdown
*[EPA]: Environmental Protection Act
*[H(PHYR) Act]: Horses (Protective Headgear for Young Riders) Act
*[H(PHYR) Regulations]: Horses (Protective Headgear for Young Riders) Regulations
*[MC(EP)R]: Motor Cycles (Eye Protectors) Regulations
*[MC(PH)R]: Motor Cycles (Protective Helmets) Regulations
*[MT(S)R]: Motorways Traffic (Scotland) Regulations
*[MT(S)(A)R]: Motorways Traffic (Scotland) (Amendment) Regulations
*[MV(DL)R]: Motor Vehicles (Driving Licences) Regulations
*[MV(VSL)(E&W)]: Motor Vehicles (Variation of Speed Limits) (England & Wales) Regulations
*[MV(WSB)R]: Motor Vehicles (Wearing of Seat Belts) Regulations
*[MV(WSB)(A)R]: Motor Vehicles (Wearing of Seat Belts) (Amendment) Regulations
*[MV(WSBCFS)R]: Motor Vehicles (Wearing of Seat Belts by Children in Front Seats) Regulations
*[NRSWA]: New Roads and Streetworks Act
*[PCUR]: Pedal Cycles (Construction & Use) Regulations
*[PCC(S)A]: Powers of Criminal Courts (Sentencing) Act
*[PRA]: Police Reform Act
*[TPSCP(S)R]: Prohibition of Smoking in Certain Premises (Scotland) Regulations
*[RSA]: Road Safety Act
*[RTA]: Road Traffic Act
*[PLSR]: Road Traffic Act 1988 (Prescribed Limit)(Scotland) Regulations
*[RT(ND)A]: Road Traffic (New Drivers) Act
*[RTOA]: Road Traffic Offenders Act
*[RV(DRM)R]: Road Vehicles (Display of Registration Marks) Regulations
*[RVLR]: Road Vehicles Lighting Regulations
*[RV(R&L)R]: Road Vehicles (Registration & Licensing) Regulations
*[TSf(EV)R]: Smoke-free (Exemptions and Vehicles) Regulations
*[TSfP(W)R]: Smoke-free Premises etc (Wales) Regulations
*[S-f(PV)R]: Smoke-free (Private Vehicles) Regulations
*[S-f(W)R]: Smoke-free Premises etc (Wales) (Amendment) Regulations
*[TMA]: Traffic Management Act
*[TSRGD]: Traffic Signs Regulations & General Directions
*[UICHR]: Use of Invalid Carriages on Highways Regulations
*[VERA]: Vehicle Excise and Registration Act
*[ZPPPCRGD]: Zebra, Pelican and Puffin Pedestrian Crossings Regulations and General Directions
*[CSDPA]: Chronically Sick & Disabled Persons Act
*[FTWO]: Functions of Traffic Wardens Order
*[GL(GP)A]: Greater London (General Powers) Act
*[HA]: Highway Act
*[MT(E&W)R]: Motorways Traffic (England & Wales) Regulations
*[MT(E&W)(A)R]: Motorways Traffic (England & Wales) Amended Regulations
*[PCUR]: Pedal Cycles (Construction & Use) Regulations
*[PPVA]: Public Passenger Vehicles Act
*[RTA]: Road Traffic Act
*[RTRA]: Road Traffic Regulation Act
*[CUR]: Road Vehicles (Construction & Use) Regulations
*[R(S)A]: Roads (Scotland) Act
```

So a lot of parens and ampersands in the labels.

I also found:

```markdown
*[TNFα]: tumour necrosis factor alpha
```

So there's an example of some non-ascii chars.

All told there are around 300 pages with abbreviations with non alphanumerics in their labels, so it's at least rare that they're used in that way.

I haven't looked so much at the titles, but the testing above shows that it won't be parsed as markdown (which is just as well, because it needs to go in an attribute, so it can't have tags in it).

**Reply by richardTowers** [reply]
A real world example of a non-ascii abbreviation (which works in kramdown), we have this Bengali abbreviation:

```
এনআইপিটি

*[এনআইপিটি]: নন-ইনভেসিভ প্রিন্যাটাল টেস্টিং
```

(aside: Bengali is a very beautiful script)

**Reply by richardTowers** [reply]
I had a bit of a go trying to use `resolveAll`, but I haven't been able to find an appropriate place for it. Seems if I put it on the `abbrDefinition` tokenizer I don't get the right events (because we're only tokenizing `contentInitial`), but it only gets run on `text` if the `tokenize` call succeeds, and there's nothing I want to tokenize in text, other than the `abbrCalls` which I'm trying to use `resolveAll` for. Does that make any sense? I'm probably a bit confused.

I did a draft PR to show what I've tried: https://github.com/richardTowers/micromark-abbr/pull/1

**Reply by wooorm** [reply]
* so, interesting that no case normalization is performed for abbreviations in Kramdown. Those do happen when matching URL references and definitions.
* numbers (at least, ascii digits) are the same as letters
* punctuation (at least, ascii, `-` and `+`) can at least be used inside word, your examples also add `(`, `)`, `&`.
* your examples also add spaces! `*[H(PHYR) Regulations]`. That’s surprising?
* non-ascii such as `α` and Bengali is fine

> Does that make any sense? I'm probably a bit confused.

No that makes sense. That’s indeed a currently imposed limitation.
There’s no way around that now. It would somehow add the resolver to https://github.com/micromark/micromark/blob/4bcb4cc58ff360b2e7db15c1f60995d806b6827f/packages/micromark/dev/lib/create-tokenizer.js#L64, but that isn’t supported yet.

It is possible to set resolvers from an extension to be called when a “span” is done (e.g., `*` and `*`, then you need to “resolve” the emphasis/links inside them). But that doesn’t support the “whole”:

```js
/**
 * @import {Resolver} from 'micromark-util-types'
 */

import {micromark} from 'micromark'

const result = micromark('FBI and *CIA*.', {
  extensions: [{insideSpan: {null: [{resolveAll: resolveAbbreviationCalls}]}}]
})

console.log(result)


/** @type {Resolver} */
function resolveAbbreviationCalls(events, context) {
  for (const [kind, token] of events) {
    if (kind === 'exit') continue
    console.log([token.type, context.sliceSerialize(token)])
  }

  return events
}
```

**Reply by richardTowers** [reply]
Great, thanks for clarifying!

I'm currently planning on shooting for a `remark-abbr` extension instead. It will add a syntax-only extension to micromark which only 
covers abbreviation definitions, and a mdast extension to transform the abbreviation calls. I'm not planning on publishing the micromark extension separately to npm, because it's not very useful on its own. As part of the unified chain though, it should work just as well as any other extension (I think?). [Work in progress on this branch](https://github.com/richardTowers/micromark-abbr/tree/syntax-only) (repo name will need to change at some point).

The `zmarkdown` folks already have a [`remark-abbr`](https://github.com/zestedesavoir/zmarkdown/tree/master/packages/remark-abbr), but it hasn't been updated to use micromark yet. If I can get my version working to an acceptable standard, I'll see if they'll accept a PR.

**Reply by richardTowers** [reply]
If a future version of micromark adds some functionality that would make an html compiler for abbreviations doable without ugly workarounds, I'd be happy to try to help get the extension written.

**Reply by wooorm** [reply]
Right, cool! Yes. I think it’s good to start with something that works. Make something. See if people want it. Most folks will use ASTs. We can see about plain micromark later! Do please let me know how it goes!
